apps/drec-api/src/pods/device/device.service.ts

```diff
@@ -107,4 +119,68 @@ export class DeviceService {
     }
     return Countries.filter((country) => country.name === countryName)[0].code;
   }
+
+  public async addToGroup(
+    currentDevice: Device,
+    groupId: number,
+    organizationOwnerCode?: string,
+  ): Promise<Device> {
+    const deviceExists = await this.getDeviceForGroup(
+      currentDevice.id,
+      groupId,
+    );
+    if (deviceExists) {
+      const message = `Device with id: ${currentDevice.id} already added to this group`;
+      this.logger.error(message);
+      throw new ConflictException({
+        success: false,
+        message,
+      });
+    }
+    if (currentDevice.groupId) {
+      const message = `Device with id: ${currentDevice.id} already belongs to a group`;
+      this.logger.error(message);
+      throw new ConflictException({
+        success: false,
+        message,
+      });
+    }
+    if (
+      organizationOwnerCode &&
+      currentDevice.registrant_organisation_code !== organizationOwnerCode
+    ) {
+      throw new NotAcceptableException(
+        `Device with id: ${currentDevice.id} belongs to a different owner`,
+      );
+    }
+    currentDevice.groupId = groupId;
+    return await this.repository.save(currentDevice);
+  }
+
+  public async removeFromGroup(
+    deviceId: number,
+    groupId: number,
+  ): Promise<Device> {
+    const currentDevice = await this.getDeviceForGroup(deviceId, groupId);
+    if (!currentDevice) {
+      throw new NotFoundException(
+        `No device found with id ${deviceId} and groupId: ${groupId}`,
+      );
+    }
+    currentDevice.groupId = null;
+    return await this.repository.save(currentDevice);
+  }
```

Comment:
Shouldn't we check if the user is the owner before removing a device from a group?

Reply:
The getDeviceForGroup method retrieves a device based on deviceId and deviceGroupId. And when I do the check (!currentDevice) - this is what I am checking. So in this case if the Device Id is wrong or it belongs to a different group, it should throw NotFound.

Reply:
Sure, we check the group ID.
But do we check if the user actually owns the group?

Reply:
This is done in the device-group service, before reaching this method.